contrib_bots/lib/giphy/giphy.py

```diff
@@ -0,0 +1,128 @@
+        # To prevent infinite loop in private message, bot will detect
+        # if the sender name is the bot name it will return false.
+        if message['type'] == 'private':
+            return client.full_name != message['sender_full_name']
+
+        # Return True if we want to (possibly) response to this message.
+        original_content = message['content']
+        is_giphy_called = (original_content.startswith('@giphy ') or
+                           message['is_mentioned'])
+
+        return is_giphy_called
+
+    def handle_message(self, message, client, state_handler):
+        bot_response = get_bot_giphy_response(message, client)
+
+        if message['type'] == 'private':
+            client.send_message(dict(
+                type='private',
+                to=message['sender_email'],
+                content=bot_response,
+            ))
+        else:
+            client.send_message(dict(
+                type='stream',
+                to=message['display_recipient'],
+                subject=message['subject'],
+                content=bot_response,
+            ))
+
+
+def get_url_gif_giphy(keyword, api_key):
+    # Return a URL for a Giphy GIF based on keywords given.
+    # In case of error, e.g. failure to fetch a GIF URL, it will
+    # return a number.
+    query = {'s': keyword,
+             'api_key': api_key}
+    try:
+        data = requests.get('http://api.giphy.com/v1/gifs/translate', params=query)
+    except requests.exceptions.ConnectionError as e:  # Usually triggered by bad connection.
+        logging.warning(e)
+        return 1
+
+    search_status = data.json()['meta']['status']
+    if search_status != 200 or data.status_code != 200:
+        return 1
+
+    try:
+        gif_url = data.json()['data']['images']['original']['url']
+    except (TypeError, KeyError):  # Usually triggered by no result in Giphy.
+        return 2
+
+    return gif_url
+
+def get_bot_giphy_response(message, client):
+    # Handle the message that called through mention.
+    if message['is_mentioned']:
+        if message['content'].startswith('@**%s** ' % (client.full_name)):
+            keyword = message['content'][len('@**%s** ' % (client.full_name)):]
+        elif message['content'].startswith('@%s ' % (client.full_name)):
+            keyword = message['content'][len('@%s ' % (client.full_name)):]
+        else:
+            return 'Please mention me first, then type the keyword.'
+    # Handle the message that called through the specified keyword.
+    elif message['content'].startswith('@giphy '):
+        keyword = message['content'][len('@giphy '):]
+    # Handle the private message.
+    elif message['type'] == 'private':
+        keyword = message['content']
+
+    gif_url = get_url_gif_giphy(keyword, giphy_api_key)
+
+    # Each error has a specific reply should "gif_url" return a number.
+    # The bot will post the appropriate message for the error.
+    if gif_url == 1:
```

Comment:
You could avoid these conditionals if you used exceptions in the place where you are checking status and content.